dev-support/docker/Dockerfile

```diff
@@ -14,47 +14,162 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# Dockerfile for installing the necessary dependencies for building Hadoop.
-# See BUILDING.txt.
-
-FROM maven:3.5-jdk-8
-
-RUN apt-get -q update && apt-get -q install --no-install-recommends -y \
-       git \
-       bats \
-       findbugs \
-       libperl-critic-perl \
-       pylint \
-       python-dateutil \
-       rsync \
-       make \
-       gcc \
-       libc6-dev \
-       ruby \
-       ruby-dev \
-       wget \
-       && \
-    gem install --no-document rake rubocop ruby-lint
-
-ENV FINDBUGS_HOME /usr
-
-####
-# Install shellcheck
-###
-RUN mkdir -p /opt/shellcheck && \
-    curl -L -s -S \
-        https://storage.googleapis.com/shellcheck/shellcheck-stable.linux.x86_64.tar.xz \
-        -o /opt/shellcheck.tar.xz && \
-    tar xJf /opt/shellcheck.tar.xz --strip-components 1 -C /opt/shellcheck && \
-    ln -s /opt/shellcheck/shellcheck /usr/bin/shellcheck && \
-    rm -f /opt/shellcheck.tar.xz
+#
+# Dockerfile used as the build and test environment for Yetus.
+#
+# Built in multiple stages so as to avoid re-downloading large binaries when
+# tweaking unrelated aspects of the image.
 
-###
-# Avoid out of memory errors in builds
-###
-ENV MAVEN_OPTS -Xmx3g
+# start with a minimal image into which we can download remote tarballs
+FROM ubuntu:18.04 AS BASE_IMAGE
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# hadolint ignore=DL3009
+RUN DEBIAN_FRONTEND=noninteractive apt-get -qq update && \
+  DEBIAN_FRONTEND=noninteractive apt-get -qq install --no-install-recommends -y \
+    ca-certificates=20180409 \
+    curl=7.58.0-2ubuntu3.8 \
+    locales=2.27-3ubuntu1
+
+RUN locale-gen en_US.UTF-8
+ENV LANG=en_US.UTF-8 LANGUAGE=en_US:en LC_ALL=en_US.UTF-8
+
+##
+# download sundry dependencies
+#
+
+FROM BASE_IMAGE AS FINDBUGS_DOWNLOAD_IMAGE
+ENV FINDBUGS_VERSION '3.0.1'
+ENV FINDBUGS_URL "https://downloads.sourceforge.net/project/findbugs/findbugs/${FINDBUGS_VERSION}/findbugs-${FINDBUGS_VERSION}.tar.gz"
+ENV FINDBUGS_SHA256 'e80e0da0c213a27504ef3188ef25f107651700ffc66433eac6a7454bbe336419'
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN curl -LfsSo /tmp/findbugs.tar.gz "${FINDBUGS_URL}" && \
```

Comment:
personal nit: long curl options much easier to maintain since most of our contributor don't have curl short flags memorized.